ipbench2/doc/ipbench.man
.TH "ipbench" "1" "Mar 2024" "Ian Wienand, Lesley Rossouw, Peter Chubb, Christopher Irving" "ipbench"
.SH "NAME"
.LP
ipbench \- Control an ipbench session
.SH "SYNTAX"
.HP
.ad l
ipbench
[\c
\fB\-c\fP <\fIclient\fP>|\fB\-\-client\fP=<\fIclient\fP> ...]
[\c
\fB\-p\fP <\fIport\fP>|\fB\-\-port\fP=<\fIport\fP>]
[\c
\fB\-t\fP <\fItest\fP>|\fB\-\-test\fP=<\fItest\fP>]
[\c
\fB\-a\fP <\fIargs\fP>|\fB\-\-test\-args\fP=<\fIargs\fP>]
[\c
\fB\-r\fP|\fB\-\-reset\fP]
[\c
\fB\-T\fP <\fIhostname\fP>|\fB\-\-test\-target\fP=<\fIhostname\fP>]
[\c
\fB\-P\fP <\fIport\fP>|\fB\-\-test\-port\fP=<\fIport\fP>]
[\c
\fB\-\-target\-test\-hostname\fP=<\fIhostname\fP>]
[\c
\fB\-\-target\-test\-port\fP=<\fIport\fP>]
[\c
\fB\-\-target\-test\fP=<\fItarget-test\fP>]
[\c
\fB\-\-target\-test\-args\fP=<\fIargs\fP>]
[\c
\fB\-\-test\-controller\-args\fP=<\fIargs\fP>]
[\c
\fB\-\-target\-test\-controller\-args\fP=<\fIargs\fP>]
[\c
\fB\-\-config\fP=<\fIconfig\fP>]
[\c
\fB\-d\fP|\fB--debug\fP]

.ad
.HP
ipbench [\fB\-h\fP|\fB\-\-help\fP]
.HP
ipbench \fB\-\-version\fP
.SH "DESCRIPTION"
.LP
The ipbench network benchmarking suite consists of three main components.
.LP
- A 'target' program running on the machine under study
.br
- Multiple 'client' programs running on other machines, which will test the target machine's performance
.br
- A 'controller', which tells the others what to do and coordinates starting and stopping them
.LP
The ipbench program fulfills the role of controller, allowing the user to specify where the clients
and target are, how they should communicate, and what tests they should run.
.LP
The ipbenchd program fulfills the role of client. An instance of ipbenchd should be running on
the relevant port on each client machine before ipbench is invoked.
.LP
The ipbenchd program can also fulfill the role of target, running on the target machine.
However, the target machine might not be running a complete operating system,
making it difficult to install ipbenchd. In this case you should implement your own
target program on the target machine.
.SH "OPTIONS"
.LP
.TP
\fB\-c\fP <\fIclient\fP>|\fB\-\-client\fP=<\fIclient\fP>
Hostname or IP on which a client ipbenchd instance will be running,
for the ipbench controller (this program) to send commands to. Repeat this argument
multiple times to specify multiple clients.
.TP
\fB\-p\fP <\fIport\fP>|\fB\-\-port\fP=<\fIport\fP>
Port on the client machines which the controller
will use to contact the clients. ipbenchd must be listening on this port
on each client.
.TP
\fB\-t\fP <\fItest\fP>|\fB\-\-test\fP=<\fItest\fP>
Name of test which the clients will all load and run.
.TP
\fB\-a\fP <\fIargs\fP>|\fB\-\-test\-args\fP=<\fIargs\fP>
Arguments which will be provided to each client, affecting how they
perform their test.
.TP
\fB\-r\fP|\fB\-\-reset\fP
Sends an ABORT message to each client from the controller, causing the
connection to close and reopen, each time the controller sets up for a test.
This might be useful if a test fails to complete properly.
.TP
\fB\-T\fP <\fIhostname\fP>|\fB\-\-test\-target\fP=<\fIhostname\fP>
Hostname or IP address of the target machine that the clients will use to
communicate with the target and perform their tests.
.TP
\fB\-P\fP <\fIport\fP>|\fB\-\-test\-port\fP=<\fIport\fP>
Port on the target machine that the clients will use to communicate with the
target and perform their tests.
The target needs to be listening on this port.
The target daemon can be ipbenchd in target mode, or another program.
.TP
\fB\-\-target\-test\-hostname\fP=<\fIhostname\fP>
Hostname or IP address of the target machine that the controller will use to
communicate with the target.
.TP
\fB\-\-target\-test\-port\fP=<\fIport\fP>
Port on the target machine that the controller will use to communicate with
the target and instruct it to run tests.
The target needs to be listening on this port.
The target daemon can be ipbenchd in target mode, or another program.
.TP
\fB\-\-target\-test\fP=<\fItarget-test\fP>
Name of test which the target will load and run. This allows monitoring aspects
of the target system while the clients test its network performance.
.TP
\fB\-\-target\-test\-args\fP=<\fIargs\fP>
Arguments which will be provided to the target, affecting how it performs any
tests it runs on itself.
.TP
\fB\-\-test\-controller\-args\fP=<\fIargs\fP>
This string is ultimately passed to the setup_controller function in the
implementation of any test which runs on the clients. See each client-side test
for what arguments are available.
.TP
\fB\-\-target\-test\-controller\-args\fP=<\fIargs\fP>
This string is ultimately passed to the setup_controller function in the
implementation of any test which runs on the target. See each target-side test
for what arguments are available.
.TP
\fB\-\-config\fP=<\fIconfig\fP>
Specify an input config file (XML based, see below).
This allows you to provide ipbench options with more flexibility
than the command line.
.TP
\fB\-d\fP|\fB--debug\fP
Enable debugging; verbose output.
.TP
\fB\-\-version\fP
Show program's version number and exit.
.TP
\fB\-h\fP|\fB\-\-help\fP
Show help information.
.SH "XML CONFIGURATION FILE"
Using the \fB--config\fP argument to ipbench, you can provide settings for your ipbench session
through an xml file instead of (or in addition to) command line arguments.
.br
More specific options override more general options. Command line arguments are the
most general and settings in the xml will always take precedence over them. Where an option
can be specified in multiple places in the xml, the most deeply nested instance
of the option is the one that is used. Not all options need be specified in the xml,
although some are required.
.br
The basic structure of an xml configuration file for ipbench should be as follows.
Which command line arguments correspond to which xml attributes are indicated.

.LP
.nf
	<ipbench>

		<!--
			'test' tag
			* name (required) : \fB--test\fP for all clients
			* args : \fB--test-args\fP for all clients
			* target : \fB--test-target\fP for all clients
			* port : \fB--test-port\fP for all clients
		-->

		<test name="" args="" port="" target="" controller_args="">

			<!--
				'client' tag
				* hostname (required) : \fB--client\fP for this client
				* port : \fB--port\fP for this client
				* test_target : \fB--test-target\fP for this client
				* test_port : \fB--test-port\fP for this client
				* test_args : \fB--test-args\fP for this client
			-->

			<client hostname="" port="" test_port="" test_args="" test_target=""/>

		</test>

		<!--
			'target_test' tag
			* name (required) : \fB--target-test\fP
			* args : \fB--target-test-args\fP
			* controller_args : \fB--target-test-controller-args\fP
		-->

		<target_test name="" args="" controller_args="">

			<!--
				'target' tag
				* hostname (required) : \fB--target-test-hostname\fP
				* port : \fB--target-test-port\fP
				* test_args : \fB--target-test-args\fP
			-->

			<target hostname="--target-test-hostname" port="--target-test-port" test_args="--target-test-args"/>
		</target_test>
	</ipbench>
.fi
.LP

For instance, below is an example of the same ipbench controller input expressed twice.
.br
The first case shows how the input could be given to ipbench as command line arguments.

.HP
.ad l
\fBipbench\fP \fB-p\fP 8036 \fB--client\fP 10.13.0.4 \fB--client\fP 10.13.0.5 \fB--client\fP 10.13.0.6 \fB--client\fP 10.13.0.7 \fB--test-target\fP=172.16.1.1 \fB--test-port\fP=1235 \fB--test\fP=latency \fB--test-args\fP="socktype=udp,bps=250000000,size=1472,warmup=0,cooldown=0,samples=50000" \fB--target-test\fP=cpu_target \fB--target-test-hostname\fP=172.16.1.1 \fB--target-test-port\fP=1236

.LP
The second case shows the contents of an xml file which will produce the same configuration if passed to ipbench.
.nf

	<ipbench>

		<test name="latency"
			args="socktype=udp,bps=250000000,size=1472,warmup=0,cooldown=0,samples=50000"
			port="1235" target="172.16.1.1">

			<client hostname="10.13.0.4" port="8036"/>
			<client hostname="10.13.0.5" port="8036"/>
			<client hostname="10.13.0.6" port="8036"/>
			<client hostname="10.13.0.7" port="8036"/>
		</test>

		<target_test name="cpu_target">
			<target hostname="172.16.1.1" port="1236"/>
		</target_test>
	</ipbench>
.fi
.LP

For more details on options in the xml configuration file, see the sample config.xml in the ipbench-controller package's documentation directory on your system.
.SH "FILES"
.LP
(none)
.SH "ENVIRONMENT VARIABLES"
.LP
(none)
.SH "EXAMPLES"
.LP
examples
.SH "AUTHORS"
.LP
ipbench by Ian Wienand <ianw@gelato.unsw.edu.au>,
.br
Lesley Rossouw <lesley.rossouw@unsw.edu.au>,
.br
Peter Chubb <peter.chubb@unsw.edu.au>.
.br

ChangeLog
2021-09-27  Peter Chubb  <peter.chubb@unsw.edu.au>

	* Update to Python3

2010-03-04  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* Depend on python-lxml to get full XPATH support for ipbench
	  controller. 
	* nfsstone fixes.  Used to dump core in ipbenchd when used.
	

2010-02-10  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* nfs_latency fixes -- now obeys similar arguments to the latency
	test (warmup, cooldown in seconds, samples=)
	* Updated manual.

2010-02-02  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* Fix compilation on Ubuntu Jaunty.

2006-08-28  Ian Wienand  <ianw@localdomain>

	* src/pymod/ipbench.c: fix up some warnings
	* src/pymod/ipbench.i.in: move around so that functions have
	definitions, fixes a few warnings.

2006-03-08  Peter Chubb  <peterc@gelato.unsw.edu.au>

	* debian/ipbench-plugins.install: Don't install libraries as part
	of this package; instead depend on libipbench package (which is
	done automatically).

2006-03-07  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/Makefile.am : respect DESTDIR

2005-06-29  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/Makefile.am : install libraries in python site-dir

2005-06-23  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/latency/microuptime* : move into 
	* src/libs/* : add microuptime into util library
	* src/tests/latency/latency.[c|h] : remove references to
	microuptime as moved into util library.

2005-06-21  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/nfs_latency/*: start a new nfs latency test, written
	by Luke Macpherson.

2005-05-05  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/ipbench.py, src/pymod/*, src/tests/* : modify API to have a
	controller setup function and phase.  This was we can pass
	arguments to the controller.
	* src/tests/latency.c : add ability to dump out client results to
	a file

2005-03-02  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/latency/latency.c : modify printf specifier for
	tick_rate
	* src/ipbench.py : add more information to execption
	errors including return code and error strings.

2005-02-24  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/cpu_target_lukem/cpu_target_lukem.c : wait on a
	condition variable to support FreeBSD.  also nice() process before
	starting usage thread.

2005-02-22  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* many: changes to bring consisentcy to types; turn on -Wall,
	-Werror; modify to build
	* configure.in: version to 2.0.2

2005-02-18  Mark Tsui <markt@cse.unsw.edu.au>

        * src /tests/latency/microuptime.h:
	* src /tests/cpu_target_lukem/cpu_target_lukem.h: added support for amd64

2005-02-18  Mark Tsui <markt@cse.unsw.edu.au>

	* src/tests/latency.c(latency_marshall): don't use size_t but int

2005-02-18  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/ipbench.c(marshall) : don't pass in uninitalised pointers
	* src/tests/wrapper/{client|target}_wrapper.c: as above
	* src/ipbenchd.py : set reuse addr to stop annoying socket in use problem

2005-01-25  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/tests/cpu_target_lukem/* : add lukem cpu usage test

2004-12-21  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* src/pymod/Makefile.am : install using DESTDIR
	* debian/* : debianisation

2004-10-21  Darren Williams  <dsw@quasar>

	* Intoduce a common ipbench2 include directory
	* Add bswap.h to new include directory, code taken
	  from latency.h

2004-10-20  Darren Williams  <dsw@gelato.unsw.edu.au>

	* Import nfsstone benchmark to src/tests/nfsstone.
	* See README in src/tests/nfsstone for details on the nfsstone benchmark.

2004-10-20  Ian Wienand  <ianw@gelato.unsw.edu.au>
	* src/ipbenchd.py : make mmap use -1 as the fd so it is portable
	* src/tests/http/Makefile.am : add to make sure make dist works
	* configure.in : update to make sure all makefile are created for make dist,
	add --enable-tests option to only build some tests.
	* src/tests/Makefile.am : change to follow conditional compliation from --enable-tests
	* src/tests/nfsstone : added by dsw, working in beta stage
	* src/pymod/Makefile.am : do not redefine libdir to install python libraries
	* src/pymod/ipbench.c : use pkglibdir for search path for plugins

2004-09-27  Ian Wienand  <ianw@gelato.unsw.edu.au>

	* restart : restart this ChangeLog for V2


ipbench2/src/ipbench.py
#!/usr/bin/env python3
"""
ipbench.py

Module encapsulating all controller logic.
"""

import sys
import re
import socket
import select
from copy import copy
from argparse import ArgumentParser
from lxml.etree import parse
import ipbench_client
import ipbench_target

OPTIONS = None
SELECTED_TEST = None


def dbprint(msg):
    """
    Debug print.

    Args:
        msg: message to print
    """
    global OPTIONS
    if OPTIONS.debug:
        sys.stderr.write(msg + "\n")


class IpBenchError(Exception):
    """
    Simple class for error representation.

    Attributes:
        value: Error value
    """

    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)


class IpbenchTestClient:
    """
    Class encapsulating logic for controller machine to spin up tests.
    """

    def __init__(self, hostname, port, test_target, test_port, client_id, test_name, test_ptr, test_args=""):
        self.hostname = hostname
        self.port = port
        self.test_port = test_port
        self.test_target = test_target
        self.test_args = test_args
        self.socket = None
        self.client_id = client_id    # Identifier for this client
        self.test_name = test_name
        self.test_ptr = test_ptr

        dbprint("[IpbenchTestClient:__init__] : client " + self.hostname +
                " port " + repr(self.port) + " test_port " + repr(self.test_port) + " test_args " + self.test_args)

    # for select()
    def fileno(self):
        """
        Return file number of opened socket to client.
        """
        return self.socket.fileno()

    def _parse_return_code(self, data):
        """
        Parse a return code. Used for telnet interface to clients.
        Return codes are of the form 123 CODE(MESSAGE)

        Params:
            data: Return code from telnet
        """
        status = {}
        status["code"] = int(data[0:3])
        fence1 = data.find("(")
        if (fence1 == -1):
            status["str"] = data[4:].strip()
            status["msg"] = ""
        else:
            fence2 = data.find(")")
            status["str"] = data[4:fence1].strip()
            status["msg"] = data[fence1+1:fence2].strip()

        dbprint("[parse_return_code] : "
                + " code=" + repr(status["code"])
                + "|str=" + status["str"]
                + "|msg=" + status["msg"]+"|")
        return status

    def parse_return_code(self):
        """
        Get and parse a return code from telnet.
        """
        data = self.socket.recv(1024).decode('utf-8')
        return self._parse_return_code(data)

    def send_command(self, cmd):
        """
        Send a command to the client over telnet.

        Params:
            cmd: Command to send
        """
        dbprint("[send_command] : " + cmd)
        self.socket.send((cmd + "\n").encode('utf-8'))

    def connect(self):
        """
        Connect to the client machine. Creates a socket and tests that
        communication works as expected.

        Raises an IpBenchError exception on failure.
        """
        global OPTIONS

        dbprint("[connect] start | hostname " + self.hostname +
                " | port " + repr(self.port) + " | ")

        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            try:
                self.socket.connect((self.hostname, self.port))
            except OSError:
                sys.stderr.write(f"Can't connect to {self.hostname} on port {repr(self.port)}"
                                 + " ... invalid host or port?\n")
                sys.exit(1)
            status = self.parse_return_code()

            if (status["code"] != 100):
                raise IpBenchError("Invalid Version Flag")

            if (OPTIONS.reset):
                self.send_command("ABORT")
                # the connection should have closed; re-open
                self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self.socket.connect((self.hostname, self.port))
                status = self.parse_return_code()
                if (status["code"] != 100):
                    raise IpBenchError("Invalid Version Flag")
            # continue
            # TODO: check versions
            self.send_command("HELLO")
            status = self.parse_return_code()
            if (status["code"] != 200):
                raise IpBenchError("HELLO to " + self.hostname + " failed (" +
                                   str(status["code"]) + " " + status["str"] + ")")
            self.send_command("LOAD " + self.test_name)
            status = self.parse_return_code()
            if (status["code"] == 421):
                raise IpBenchError("LOAD to " + self.hostname +
                                   " failed (" + status["msg"] + ")")
            elif (status["code"] != 200):
                raise IpBenchError("LOAD Failed")
        except IpBenchError:
            raise

    def setup(self):
        """
        Set up a client. Assumes that connection is already established.
        """

        global OPTIONS

        try:
            if self.test_port is None:
                self.test_port = 0
            if self.test_target is None:
                raise IpBenchError("Must specify --test-target")
            if self.test_args is None:
                self.test_args = ""

            dbprint("[Setup]: target: " + self.test_target)
            dbprint("[Setup]: port: " + str(self.test_port))
            dbprint("[Setup]: args: " + self.test_args)
            self.send_command("SETUP target::" + self.test_target + "||" +
                              "port::" + repr(self.test_port) + "||"
                              'args::"' + self.test_args + '"')
            status = self.parse_return_code()
            if (status["code"] != 200):
                raise IpBenchError("SETUP to " + self.hostname +
                                   "failed (" + str(status["code"]) + " " + status["str"] + ")")
        except IpBenchError:
            raise

    def start(self):
        """
        Send command to start a client.
        """
        self.send_command("START")

    def call_test_unmarshall(self, client_data, valid):
        """
        Invoke unmarshall feature in test module.
        """
        self.test_ptr.unmarshall(int(self.client_id), client_data, valid)

    def unmarshall(self):
        """
        Start an unmarshalling operation on the remote client and
        try to interpret data.
        """
        dbprint("[unmarshall] client " + self.hostname + " unmarshall ")
        buffer = self.socket.recv(128)
        statusline = copy(buffer).decode('utf-8', errors='ignore')
        newline = statusline.find('\n')
        status = self._parse_return_code(statusline[:newline])
        buffer = buffer[newline+1:]
        if (status["code"] == 220):
            valid = True
        elif (status["code"] == 221):
            valid = False
        else:
            raise IpBenchError("Test failed  : " + status["msg"])

        # clenline is the line with the content-length in it
        initial_content = buffer + self.socket.recv(128)
        clenline = copy(initial_content).decode('ascii', errors='ignore')
        newline = clenline.find('\n')
        r = re.match(r"Content-length: (\d+)", clenline[:newline])
        if r.group(1) is None:
            raise IpBenchError("Invalid Content-length")
        content_length = int(r.group(1))
        dbprint("[unmarshall] client " + self.hostname +
                " content-length: " + repr(content_length))

        client_data = initial_content[newline + 1:]
        while (len(client_data) != content_length):
            dbprint("[unmarshall] client " + self.hostname +
                    " len(" + repr(len(client_data)) + ") < " + repr(content_length))
            newdata = self.socket.recv(content_length - len(client_data))
            dbprint(("[unmarshall] got " + repr(len(newdata)) + " new bytes "))
            client_data = client_data + newdata

        # wrapper, as the target test doesn't need to be passed an id
        self.call_test_unmarshall(client_data, valid)

    def close(self):
        """
        Terminate a connection with the client. Triggers the ipbench daemon to exit
        and closes socket.
        """
        self.send_command("QUIT")
        self.socket.close()


# the test target is very similar, however it doesn't have the test_target
# hostname since it has no target!
class IpbenchTestTarget(IpbenchTestClient):
    """
    Class encapsulating control over an ipbench target instance for the controller.

    This is only used when a test which expects an instance of ipbenchd --target
    to be running is used. All logic here is simply to manipulate that ipbenchd instance.
    """

    def __init__(self, hostname, port, test_name, test_ptr, test_args=""):
        self.hostname = hostname
        self.port = port
        self.s = None
        self.test_args = test_args
        self.test_name = test_name
        self.test_ptr = test_ptr
        dbprint("[IpbenchTestTarget:__init__] : client " + hostname)
        # TODO: fix inheritance here. Bad practice to have this inherit but never call super or do much with it.

    # setup needs to be slightly different because it only sends the
    # arguments
    def setup(self):
        global OPTIONS

        try:
            self.send_command('SETUP args::"' + self.test_args + '"')
            status = self.parse_return_code()
            if status["code"] != 200:
                raise IpBenchError(
                    "SETUP Failed (" + str(status["code"]) + " " + status["str"] + ")")
        except IpBenchError:
            raise

    # clients don't have a stop command
    def stop(self):
        """
        Trigger test target to stop.
        """
        self.send_command("STOP")

    def call_test_unmarshall(self, client_data, valid):
        """
        Begin unmarshalling operation on test module.
        """
        self.test_ptr.unmarshall(client_data, valid)


def error_exit(error_msg):
    print(error_msg)
    sys.exit(1)


##
# main function
##


def main():
    """
    Main function for ipbench tester.
    """
    global OPTIONS
    clients = []
    targets = []

    parser = ArgumentParser(description="ipbench is the controller program for the ipbench benchmarking suite. It communicates with clients and targets and instructs them to run tests.")
    parser.add_argument("--version", action="version", version="%(prog)s 2.1.1")
    parser.add_argument("-d", "--debug", dest="debug", action="store_true",
                      help="Enable debugging; verbose output.", default=False)

    parser.add_argument("-c", "--client", dest="clients", action="append",
                      help="Hostname or IP of a client for the controller to connect to. Repeat this argument multiple times to specify multiple clients.", default=[])

    parser.add_argument("-p", "--port", dest="port",
                      help="Port on the clients which the controller will connect to.",
                      type=int, default=8036, action="store")

    parser.add_argument("-t", "--test", dest="test", action="store",
                      help="Name of test the clients will run.", default=None)

    parser.add_argument("-a", "--test-args", dest="test_args", action="store",
                      type=str, help="Arguments affecting how the clients run their test.", default="")

    parser.add_argument("-r", "--reset", dest="reset", action="store_true",
                      help="Send an ABORT to clients before each setup. Useful for cleaning up failed tests.", default=False)

    parser.add_argument("-T", "--test-target", dest="test_target", action="store",
                      help="Hostname or IP of the target that the clients will direct their tests at.",
                      default=None)

    parser.add_argument("-P", "--test-port", dest="test_port", action="store", type=int,
                      help="Port on the target that the clients will direct their tests at.",
                      default=0)

    parser.add_argument("--target-test-hostname", dest="target_test_hostname", action="store", type=str,
                      help="Hostname or IP of the target for the controller to connect to.", default=None)

    parser.add_argument("--target-test-port", dest="target_test_port", action="store", type=int,
                      help="Port on the target that the controller will connect to.",
                      default=8037)

    parser.add_argument("--target-test", dest="target_test", action="store", type=str,
                      help="Name of test the target will run.", default=None)

    parser.add_argument("--target-test-args", dest="target_test_args", action="store", type=str,
                      help="Arguments affecting how the target runs its test.", default="")

    parser.add_argument("--test-controller-args", dest="controller_args", action="store", type=str,
                      help="Arguments for the setup_controller function of the test running on the clients. See individual tests for details.",
                      default=None)

    parser.add_argument("--target-test-controller-args", dest="target_controller_args", action="store", type=str,
                      help="Arguments for the setup_controller function of the test running on the target. See individual tests for details.",
                      default=None)

    parser.add_argument("--config", dest="config", action="store",
                      help="Specify an input config file (XML based, see documentation). Allows flexible and repeatable configuration without command line arguments.", default=None)

    OPTIONS = parser.parse_args()

    # read in a config file, if specified, and build up a list of clients
    if OPTIONS.config:
        dbprint("Reading config from " + OPTIONS.config)
        doc = parse(OPTIONS.config).getroot()

        tests = doc.findall('test')
        if (len(tests) > 1):
            error_exit("Error in config: please only have one <test> section")

        # go through attributes for <test>
        for test in tests:
            OPTIONS.test = str(test.get("name", None))
            OPTIONS.test_args = str(test.get('args', ''))
            OPTIONS.test_port = int(test.get('port', 0))
            OPTIONS.test_target = str(test.get('target', None))
            OPTIONS.controller_args = str(test.get('controller_args', None))

        # the test should be known by now
        if OPTIONS.test is None:
            error_exit("Error in config: please specify a test with either --test or in <test>")

        # go through each <client>, override default values from attributes and add it to the
        # clients[] list
        data = tests[0].findall('./client')
        for cclients in data:
            newclient = {
                "hostname": None,
                "port": OPTIONS.port,
                "test_port": OPTIONS.test_port,
                # careful with encodings
                "test_args": str(OPTIONS.test_args),
                "test_target": str(OPTIONS.test_target)
            }
            newclient["hostname"] = cclients.get('hostname', None)
            newclient["port"] = int(cclients.get('port', OPTIONS.port))
            newclient["test_port"] = int(
                cclients.get('test_port', OPTIONS.test_port))
            newclient["test_args"] = cclients.get(
                'test_args', str(OPTIONS.test_args))
            newclient["test_target"] = cclients.get(
                'test_target', str(OPTIONS.test_target))

            if newclient["hostname"] is None:
                error_exit("Please specifiy a hostname for the client!")
            clients.append(newclient)

        # now look for the <target_test> section, add to the targets list
        target_tests = doc.findall("target_test")
        for target_test in target_tests:
            target_test_name = target_test.get('name')
            OPTIONS.target_test_args = target_test.get('args', '')
            OPTIONS.target_controller_args = str(
                target_test.get('controller_args'))

            # .get returns None if not found
            if target_test_name == None:
                error_exit("Please specify a name for the <target_test>!")

            # it is possible that there will be multiple <target_test>
            # sections, each  with multiple <targets> in them.  so
            # make sure we only select those targets for the target
            # test we are currently looking at.
            tclients = doc.findall(
                "target_test[@name=\'"+target_test_name+"\']/target")
            if (not (len(tclients) > 0)):
                error_exit("Please specify some targets in the <target_test> section!")
            for tclient in tclients:
                newtarget = {
                    "hostname": OPTIONS.target_test_hostname,
                    "port": OPTIONS.target_test_port,
                    # workaround -- by default
                    "test_args": str(OPTIONS.target_test_args),
                    "test": str(target_test_name)  # comes encoded in utf-16
                }
                for arg in list(tclient.keys()):
                    newtarget["hostname"] = tclient.get('hostname',
                                                        OPTIONS.target_test_hostname)
                    newtarget["port"] = int(tclient.get(
                        'port', OPTIONS.target_test_port))
                    newtarget["test_args"] = str(tclient.get(
                        'test_args', OPTIONS.target_test_args))

                    if newtarget["hostname"] is None:
                        error_exit("Please specify a hostname for the target!")

                targets.append(newtarget)

    # finally, append things that are specified from the command line
    if OPTIONS.clients:
        if OPTIONS.test is None:
            error_exit("Please specify a test with --test!")
        for client in OPTIONS.clients:
            newclient = {
                "hostname": client,
                "port": OPTIONS.port,
                "test_port": OPTIONS.test_port,
                "test_args": OPTIONS.test_args,
                "test_target": OPTIONS.test_target
            }
            clients.append(newclient)

    # you are more limited in your target test choices from the
    # command line, for example you can only specify one target
    # and one test.  but the simple case is covered
    if OPTIONS.target_test:
        if OPTIONS.target_test_hostname is None:
            error_exit("Please specify a target test hostname with --target-test-hostname!")
        newtarget = {
            "hostname": OPTIONS.target_test_hostname,
            "port": OPTIONS.target_test_port,
            "test_args": OPTIONS.target_test_args,
            "test": OPTIONS.target_test
        }
        targets.append(newtarget)

    # sanity check stuff
    if OPTIONS.test is None:
        error_exit("Please specify a test")

    if len(clients) == 0:
        error_exit("Please specify some clients")

    # seeing as all clients are running the same test, we
    # only need one object to interact with them.
    client_test = ipbench_client
    if OPTIONS.debug:
        client_test.enable_debug()
    client_test.load_plugin(OPTIONS.test)
    client_test.setup_controller(len(clients), OPTIONS.controller_args)

    try:
        # setup each of the targets
        for target in targets:
            dbprint("[main] setting up target " + target["hostname"])

            target["shuntobj"] = ipbench_target
            if OPTIONS.debug:
                target["shuntobj"].enable_debug()
            target["shuntobj"].load_plugin(target["test"])
            target["shuntobj"].setup_controller(OPTIONS.target_controller_args)

            target["testobj"] = IpbenchTestTarget(target["hostname"], target["port"],
                                                  target["test"], target["shuntobj"],
                                                  test_args=target["test_args"])
            target["testobj"].connect()
            target["testobj"].setup()

        # setup each of the clients.
        client_id = 0
        for client in clients:
            dbprint("[main] setting up client " + client["hostname"])

            if client["test_port"] == 0:
                client["test_port"] = client_test.get_default_port()

            client["testobj"] = IpbenchTestClient(client["hostname"], client["port"],
                                                  client["test_target"], client["test_port"],
                                                  client_id, OPTIONS.test, client_test,
                                                  test_args=client["test_args"])
            client["testobj"].connect()
            client_id = client_id + 1
            client["testobj"].setup()

        # start the target test(s) first (wait for an OK response)
        for target in targets:
            target["testobj"].start()

        # send START command (error responses will be polled for below)
        for client in clients:
            dbprint("[main] client " + client["hostname"] + " start")
            client["testobj"].start()

        wait_to_read = []
        for client in clients:
            wait_to_read.append(client["testobj"])
        wait_for_error = []
        for target in targets:
            wait_for_error.append(target["testobj"])

        # simply select() on all the fd's of the clients
        # (client["testobj"].fileno() provides the fd)
        # once they have data, unmarshall it.
        while len(wait_to_read) > 0:
            (to_read, to_write, to_except) = select.select(
                wait_to_read, [], [], 0.2)
            for client in to_read:
                client.unmarshall()
                wait_to_read.remove(client)
            # additionally; poll any targets incase they have returned
            # some sort of error and we need to abort.  don't wait around
            # for this however
            (to_read, to_write, to_except) = select.select(
                wait_for_error, [], [], 0.2)
            if len(to_read) > 0:
                raise IpBenchError("Target returned an error!")

        # once all clients have reported in, stop any target tests
        # and get their data
        for target in targets:
            target["testobj"].stop()
            target["testobj"].unmarshall()

        # all the clients aggregate their data and output() shows it
        client_test.output()

        # each target gets to display it's own output
        for target in targets:
            target["shuntobj"].output()

    except IpBenchError as err:
        print("Ipbench failed : " + err.value)

    # close all connections and quit
    for client in clients + targets:
        client["testobj"].close()


if __name__ == "__main__":
    main()
